conf/parse.go

```diff
@@ -98,13 +102,42 @@ func ParseFileWithChecks(fp string) (map[string]interface{}, error) {
 	return p.mapping, nil
 }
 
+// ParseWithDigest returns the processed config and a digest
+// that represents the configuration.
+func ParseWithDigest(fp string) (map[string]interface{}, string, error) {
+	data, err := os.ReadFile(fp)
+	if err != nil {
+		return nil, _EMPTY_, err
+	}
+	p, err := parse(string(data), fp, true)
+	if err != nil {
+		return nil, _EMPTY_, err
+	}
+	digest := sha256.New()
+	e := json.NewEncoder(digest)
+	err = e.Encode(p.mapping)
+	if err != nil {
+		return nil, _EMPTY_, err
+	}
+	return p.mapping, fmt.Sprintf("sha256:%x", digest.Sum(nil)), nil
+}
+
 type token struct {
 	item         item
 	value        interface{}
 	usedVariable bool
 	sourceFile   string
 }
 
+func (t *token) MarshalJSON() ([]byte, error) {
+	// Do not duplicate variable contents since
+	// they already have been expanded at this point.
+	if t.usedVariable {
+		return []byte("null"), nil
```

Comment:
What is []byte("null") here?

Reply:
it is so that the environment variables that have been already expanded are not duplicated, so with this for example:

```hcl
BAR_USERS = [{user: a}]

accounts {
  A { users = $BAR_USERS }
  B { users = [{user: b}]}
}
```

when expanded it becomes in the map:

```hcl
{"BAR_USERS":[{"user":"a"}],"accounts":{"A":{"users":[{"user":"a"}]},"B":{"users":[{"user":"b"}]}}}
```

so added this check to skip that with a `null` value when creating the expanded version:

```
{"BAR_USERS":null,"accounts":{"A":{"users":[{"user":"a"}]},"B":{"users":[{"user":"b"}]}}}
```

the right way though would be to skip including the env vars into the result after they have been expanded, not just leaving them undefined...

Reply:
Feels brittle to me, what do you think? Should we try to remove them and just leave expanded? If we ship with them in we can never remove them and have the hash match.

Reply:
Updated to clear the env vars before taking the digest so they would now match:

```go
		{
			`# substitutions
                        BAR_USERS = { users = [ {user = "bar"} ]}
                        hello = 'world'
                        accounts {
                          QUUX_USERS = [ { user: quux }]
                          bar = $BAR_USERS
                          quux = { users = $QUUX_USERS }
                        }
                        very { nested { env { VAR = 'NESTED', quux = $VAR }}}
                        `,
			nil,
			"sha256:34f8faf3f269fe7509edc4742f20c8c4a7ad51fe21f8b361764314b533ac3ab5",
		},
		{
			`# substitutions, same as previous one without env vars.
                        hello = 'world'
                        accounts {
                          bar  = { users = [ { user = "bar" } ]}
                          quux = { users = [ { user: quux   } ]}
                        }
                        very { nested { env { quux = 'NESTED' }}}
                        `,
			nil,
			"sha256:34f8faf3f269fe7509edc4742f20c8c4a7ad51fe21f8b361764314b533ac3ab5",
		},
```